[Downey] Step 1 - Date포멧, 클래스 계층

dahun-lee-daji opened the pull request:
- 작업 목록
- [x] 클래스 계층 설계
- [x] Date

- 키워드 : class, protocol, Date, DateFormatter, DateComponentsFormatter

- 고민과 해결 : 
   - Date -> String, String-> Date, Date를 second로 바꾸어 (보통 timeInterval) 원하는 형식으로 반환 할 수 있다는 것을 이전 학습을 통해 알 고 있었으므로, 복습했습니다.
   - 클래스의 계층을 어떻게 설계 할 지 고민이 많았습니다. 
   1. 초기 : protocol 선언, Milk 등이 이를 채택 -> 중복 코드 다수
   2. 수정 1 : 초기에 선언한 protocol을 class로 변환 -> protocol을 사용하여 규격화와 추후 혹시 모를 재사용성 확보를 원함
   3. 결과 : protocol 선언 -> 해당 protocol을 채택한 class Drink 선언 -> Drink를 상속한 하위 class (Milk, Cola ....)

dahun-lee-daji committed:
Init project

dahun-lee-daji committed:
불필요한 코드 삭제

dahun-lee-daji committed:
model 구현, test출력 구현

DrinkProtocol 구현
class Drink : DrinkProtocol 구현
Drink를 상속한 Milk, Soda, Coffee 구현

godrm commented on VendingMachineApp/VendingMachineApp/AppDelegate.swift:
```diff
@@ -0,0 +1,30 @@
+
+    func application(_ application: UIApplication, configurationForConnecting connectingSceneSession: UISceneSession, options: UIScene.ConnectionOptions) -> UISceneConfiguration {
+        // Called when a new scene session is being created.
+        // Use this method to select a configuration to create the new scene with.
```
이전 단계에서 했던 피드백을 최대한 적용해보세요.  
프로젝트 템플릿으로 자동 생성되는 함수나 주석 등은 직접 사용하지 않는 경우는 과감하게 지우셔도 됩니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drink.swift:
```diff
@@ -0,0 +1,37 @@
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMdd"
+        return dateFormatter.string(for: self) ?? "00000000"
+    }
```
Date를 확장한 거면 별도 파일로 분리하세요. 여기는 음료수 타입을 위한 파일이니까요
convertDate2yyyymmddString() 이름이 길고 어색하네요. format까지 다 이름에 넣는게 좋은걸까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drink.swift:
```diff
@@ -0,0 +1,37 @@
+        self.volume = volume
+        self.charge = charge
+        self.name = name
+        self.manufacturingDate = manufacturingDate
```
어차피 타입이 Date면 속성 이름에 Date를 붙일 필요는 없습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/DrinkProtocol.swift:
```diff
@@ -0,0 +1,17 @@
+    var volume : Int {get }
+    var charge : Int {get }
+    var name : String {get }
+    var manufacturingDate : Date {get }
```
Drink 가 최상위 클래스면 DrinkProtocol 프로토콜이 아니더라도 상위 클래스 역할을 하게 됩니다. 
중복되지는 않는지 고민해보세요.

dahun-lee-daji replied:
저도 중복된다고 생각합니다. 그래서 둘 중 어떤 것을 사용 할 지 고민해 봤습니다. 
코드 중복을 줄이기 위해 상속을 사용하려고 합니다. 수정 후 다시 코멘트 드리겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Milk.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+//http://emart.ssg.com/item/itemView.ssg?itemId=0000007095603&siteNo=6001&salestrNo=2034
```
이 주석은 뭘까요? 코드를 설명하는 것이 맞나요?

dahun-lee-daji replied:
아 하위 클래스의 인터넷 주소 주석은 해당 상품 페이지입니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/Milk.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+//http://emart.ssg.com/item/itemView.ssg?itemId=0000007095603&siteNo=6001&salestrNo=2034
+class Milk: Drink {
```
하위 클래스를 선언할 꺼면 꼭 수퍼 클래스 속성과 달리 하위 클래스에 별도 속성이 있어야 합니다.
그렇지 않으면 상위-하위가 구분이 되지 않는거죠.

godrm commented on VendingMachineApp/VendingMachineApp/View/OutputView.swift:
```diff
@@ -0,0 +1,16 @@
+        print(Milk.init())
+        print(Soda.init())
+        print(Coffee.init())
+    }
```
OutputView를 만드는 시도는 매우 좋습니다. 
다만 UIKit을 연습하고 있으니 UIView를 상속해서 만들어 보는 것도 괜찮은 시도 같네요.

dahun-lee-daji replied:
차후 미션에서 View에 관한 내용이 나올 때 진행하겠습니다!

godrm reviewed (changes requested):
다음 피드백을 확인해보세요. 하위 클래스 관련해서는 다음 단계에서 개선해도 되는데 다른 부분은 확인해보고 넘어갑시다.

dahun-lee-daji committed:
피드백 반영

주의 : 커밋 파편화에 더 신경써야 함.
불필요한 코드 제거
Date extention을 위한 파일 분리
DrinkProtocol 참조파일에서 제거.
DrinkProtocol 제거에 따른 Drink 수정

dahun-lee-daji commented:
수정 완료했습니다!

godrm commented on VendingMachineApp/VendingMachineApp/extention/Date.swift:
```diff
@@ -0,0 +1,16 @@
+    func convertToString() -> String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMdd"
+        return dateFormatter.string(for: self) ?? "00000000"
```
리턴값 중에 디폴트값을 결정할 때는 조금 신중할 필요가 있습니다. 
지금은 "00000000" 안쓸 것 같지만 아주 가끔씩은 그 값 자체가 필요한 경우가 생길 수 있습니다.
그래서 이런 경우는 옵셔널도 고려해볼 만 한 것 같습니다. 아니면 throw도 괜찮구요. 
오히려 비정상적인 값을 넘기는게 더 안 좋은 경우가 있습니다. 참고하세요.
